proposed/0000-UDP-hole-punching.md

```diff
@@ -0,0 +1,192 @@
+- Feature Name: UDP-hole-punching
+- Type new: feature
+- Related components: crust
+- Start Date: 22-09-2015
+- RFC PR: (leave this empty)
+- Issue number: (leave this empty)
+
+# Summary
+
+In this text we describe an implementation of UDP hole punching process
+to allow for P2P connections of nodes behind NATs in a fully decentralised
+network.
+
+# Motivation
+
+Most users of the Crust library are expected to run the software on home PCs connected
+to the internet through a router. Routers usually implement some kind of a firewall to
+prevent anyone from the outside internet to connect to user's PC directly. As Crust
+is a library that concentrates at P2P connections, it should try available techniques
+to circumvent these firewall restrictions and UDP hole punching is one such
+technique.
+
+Additionally, traditional UDP hole punching process involves a well known
+server to tell each peer's their external UDP endpoints. Since our intent
+is a fully decentralised system, this text also describes how nodes to which
+we're already connected can be used as a replacement for such server.
+
+In the following text the reader is expected to know basic NAT types and how they work.
+The section 'Methods of translation' on [wikipedia](https://en.wikipedia.org/wiki/Network_address_translation)
+gives a nice overview.
+
+# Detailed design
+
+The UDP hole punching process described here consists of two steps:
+
+1. Finding an external mapping of a UDP socket
+2. Do the actual hole punching.
+
+## Finding an external mapping of a UDP socket
+
+Suppose, we have two nodes `A` and `B`, where each one is possibly behind a NAT and
+they want to connect to each other.  Routing needs to send `A` an endpoint of `B` and
+vice versa. But `A` and `B` don’t know their public
+endpoints a priori (they may know their IP addresses from previous connection establishments,
+but not their ports). Each node therefore needs to create a UDP socket and use it to ask some
+other node it is connected to what the public endpoints of the UDP socket is. We’ll refer
+to such helping nodes `C(A)` and `C(B)` (it needs not be the same node for `A` and `B`).
+We'll call `U(C(X))` the UDP port on which X can contact `C(X)` in order to find its public
+endpoint. `X` finds out about `U(C(X))` through the initial handhsake, that is, when
+`X` connects to `C(X)`.
```

Comment:
> It'd contain the public IP discovered. For the mapped IGD address, the IGD reported address.
So it would be the external address of the IGD device, no? In most cases this would be the same as what `peer_addr` will give us if we're already connected to `C(X)`. The only case where this isn't true is if `C(X)` had two interfaces, and each of these interfaces was connected to a different gateway (NAT). I think this is really uncommon for PCs. If it was just a simple addition which would make the code more robust but wouldn't make the other cases more complicated then I would like it, but with my current understanding I think it will just complicate the most common cases.

For example, if we have the non-IGD version, you suggest to use IP reported by other peers, but that sounds like a "chicken and an egg" thing, that is, for other peer to tell you your address, you must first know his address (so that you can ping him so that he can ping you back with your external address in the message).

Another problem: in case of IGD, there is another scenario where one could have one IGD device behind another NAT, in such case `peer_addr` would give better result.

And last that comes to my mind: If no IGD device has responded yet, and no one told us our external endpoint, what IP do we send back?

I think, in general, sooner or later we should implement your idea, it's just that I think we don't have all corner cases sorted yet, so I would prefer to keep things simple before we do.